staging/kos/pkg/networkfabric/factory.go

```diff
@@ -0,0 +1,66 @@
+/*
+Copyright 2018 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package networkfabric
+
+import (
+	"fmt"
+	"strings"
+)
+
+// netFabricFactory is the signature of
+// the factories of network fabrics.
+type netFabricFactory func() Interface
+
+var factoryRegistry map[string]netFabricFactory
+
+// registerFactory registers a netFabricFactory under the given name.
+// After a successful invocation, the registered factory can
+// be used to get an instance of a network fabric by invoking
+// NewNetFabricForName using the same name used for registration
+// as the input parameter. This function is meant to be used
+// ONLY in init() functions of network fabric implementers.
+// Invoking registerFactory more than once with the same name panics.
+func registerFactory(name string, factory netFabricFactory) {
```

Comment:
Why does the registry hold factories instead of `Interface`s?

Reply:
Fixed in d6e6daf. I used factories because a fabric could be associated with resources (connections, etc...). If more than one fabric type is registered, you are wasting resources allocated to fabrics not being used. Registering factories was my way to solve the problem: you only register a function, only the resources which are needed are allocated and this happens when you actually need them. But I agree with you on directly registering `Interface`s, as it's more explicit, in our case we have only two types of fabrics and one is lightweight (logger), and if the problem of resources allocation arises we can solve it by adding a `Run()` or `Activate()` method to `Interface` which acquires the resources the fabric need, and have the user invoke it before usage.

Reply:
Actually that's a good reason to use factories.  Using a factory is cleaner than bothering the consumer with a `Start` method.

It would be best if we could run the connection agent with the logger fabric in a context where the OVS fabric will not even start (e.g., where OVS is not installed).